src/NHibernate/AdoNet/ConnectionManager.cs

```diff
@@ -63,6 +68,16 @@ public interface Callback
 			ownConnection = suppliedConnection == null;
 		}
 
+		public void AddSessionSharingManager(ISessionImplementor session)
+		{
+			_sessionsSharingManager.Add(session);
+		}
+
+		public void RemoveSessionSharingManager(ISessionImplementor session)
+		{
+			_sessionsSharingManager.Remove(session);
```

Comment:
`O(n)`, true, but well, we are not supposed to open a lot of sessions from a single one. Otherwise, we would have to use the `OrderedHashSet` from Iesi, since we need to retain the ordering.